File: validation/poc-00/java-candidate/src/test/java/dev/vra/poc00/ArchitectureTest.java
package dev.vra.poc00;

import com.tngtech.archunit.core.importer.ClassFileImporter;
import com.tngtech.archunit.core.importer.ImportOption;
import org.junit.jupiter.api.Test;
import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses;

class ArchitectureTest {
    @Test void dependenciesPointInward() {
        var classes = new ClassFileImporter().withImportOption(new ImportOption.DoNotIncludeTests())
                .importPackages("dev.vra.poc00");
        noClasses().that().resideInAPackage("..domain..").should().dependOnClassesThat()
                .resideInAnyPackage("org.springframework..", "jakarta.servlet..", "java.sql..",
                        "..interfaces..", "..infrastructure..", "..application..").check(classes);
        noClasses().that().resideInAnyPackage("..application..", "..infrastructure..")
                .should().dependOnClassesThat().resideInAPackage("..interfaces..").check(classes);
        noClasses().that().resideInAPackage("..application..")
                .should().dependOnClassesThat().resideInAPackage("..infrastructure..").check(classes);
        noClasses().that().resideInAPackage("..interfaces..")
                .should().dependOnClassesThat().resideInAPackage("..infrastructure..").check(classes);
    }
}

File: validation/poc-00/kotlin-candidate/src/test/kotlin/dev/vra/poc00/kotlin/ArchitectureTest.kt
package dev.vra.poc00.kotlin

import com.tngtech.archunit.core.domain.JavaClasses
import com.tngtech.archunit.core.importer.ClassFileImporter
import com.tngtech.archunit.core.importer.ImportOption
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ArchitectureTest {
    @Test
    fun kotlinProductionPackagesPointInward() {
        val classes: JavaClasses = ClassFileImporter()
            .withImportOption(ImportOption.DoNotIncludeTests())
            .importPackages("dev.vra.poc00.kotlin")

        // Prove that the bytecode importer saw Kotlin classes in the intended package tree.
        assertThat(classes.any { it.name == "dev.vra.poc00.kotlin.domain.Money" }).isTrue()
        assertThat(classes.any { it.name == "dev.vra.poc00.kotlin.application.ReservationApplicationService" }).isTrue()

        noClasses().that().resideInAPackage("..kotlin.domain..")
            .should().dependOnClassesThat()
            .resideInAnyPackage(
                "org.springframework..", "jakarta.servlet..", "java.sql..",
                "..kotlin.interfaces..", "..kotlin.infrastructure..", "..kotlin.application.."
            ).check(classes)

        noClasses().that().resideInAnyPackage("..kotlin.application..", "..kotlin.infrastructure..")
            .should().dependOnClassesThat().resideInAPackage("..kotlin.interfaces..").check(classes)
        noClasses().that().resideInAPackage("..kotlin.application..")
            .should().dependOnClassesThat().resideInAPackage("..kotlin.infrastructure..").check(classes)
        noClasses().that().resideInAPackage("..kotlin.interfaces..")
            .should().dependOnClassesThat().resideInAPackage("..kotlin.infrastructure..").check(classes)
    }
}
